content/inside-rust/call-for-testing-hint-mostly-unused.md

```diff
@@ -0,0 +1,207 @@
+minutes to compile, and when you use these large crates as dependencies, they
+can take a disproportionate amount of the entire compilation time of your
+top-level crate.
+
+In some cases, crates add feature flags to control compilation of their API
+surface. This can improve compile time, but adds complexity for users, who now
+need to determine which features they need for the APIs they use. Features also
+constitute a stable interface of a crate, and changing feature flags can be a
+breaking change. And even with feature flags, not every enabled function will
+be needed; there is a balance between granularity and ease of use.
+
+## Deferring code generation with `-Zhint-mostly-unused`
+
+The latest nightly `rustc` compiler now supports an option
+`-Zhint-mostly-unused`, which tells `rustc` that the crate's API surface will
+mostly go unused. This is a hint, and `rustc` doesn't make guarantees about its
+exact behavior (so that we can extend or improve it in the future), but
+currently it causes the compiler to defer as much of code generation as
+possible.
+
+Applying this option to key crates you depend on (and use only a small subset
+of) can provide a substantial reduction in compile time, for debug builds and
+especially for release builds.
+
+## How does this perform?
+
+Some build timings for clean release builds of a crate depending on various
+specific large API crates:
+
+| **Dependency Crate** | **Before** | **`hint-mostly-unused`** | **Delta** |
+| :- | -: | -: | -: |
+| `windows`, all Graphics/UI features | 18.3s | 10.7s | -42% |
+| `windows`, all features | 3m 48s | 2m 55s | -23% |
+| `rustix`, `all-apis` feature | 5.9s | 4.3s | -27% |
+| `x11rb` and `x11rb-protocol` | 5.3s | 2.6s | -51% |
+| `aws-sdk-ec2` | 4m 07s | 2m 04s | -50% |
+
+This performance improvement comes from deferring code generation. For
+instance, the `windows` crate in the first example goes from building in 15.1s
+of which 49% is codegen, to building in 7.5s of which 1% is codegen.
+
+Note that this option does not provide a universal performance improvement for
+every crate; if used when not applicable, this option can make builds much
+*slower*. Deferring compilation of the items in a crate can lead to redoing
+code generation for those items repeatedly. In particular, avoid using this
+hint for crates whose API surface is mostly used, and/or used in multiple
+different crates or binaries (e.g. multiple test binaries that each test a
+substantial swath of the API).
+
+Always do performance analysis when considering this hint, and only apply it if
+it applies obvious and substantial wins for your users. Never apply it across
+the board to all your dependencies.
+
+If most of the items in your crate are polymorphic (generic), this hint may be
+redundant, as Rust already defers compilation of polymorphic items until they
+get monomorphized with specific types.
+
+Also note that this only provides a performance win if you are rebuilding the
```

Comment:
Is this true that it only helps when you are _rebuilding_ the dependency, or does it help whenever you are _building_ the dependency (ie even the first time)?